opm/material/common/XYTabulated2DFunction.hpp

```diff
@@ -0,0 +1,301 @@
+
+        Valgrind::CheckDefined(s1);
+        Valgrind::CheckDefined(s2);
+
+        // ... and combine them using the x position
+        result = s1 * (1.0 - alpha) + s2 * alpha;
+        Valgrind::CheckDefined(result);
+    }
+
+private:
+    // the sampling points in the x-drection
+    std::vector<Scalar> xPos_;
+    // the sampling points in the y-drection
+    std::vector<Scalar> yPos_;
+    // data at the sampling points
+    std::vector<std::vector<Scalar> > samples_;
+
+    bool xExtrapolate_ = false;
+    bool yExtrapolate_ = false;
+
+    /*!
+     * \brief Returns the minimum of the X coordinate of the sampling points.
+     */
+    Scalar xMin() const
+    { return xPos_.front(); }
+
+    /*!
+     * \brief Returns the maximum of the X coordinate of the sampling points.
+     */
+    Scalar xMax() const
+    { return xPos_.back(); }
+
+    /*!
+     * \brief Returns the minimum of the Y coordinate of the sampling points.
+     */
+    Scalar yMin() const
+    { return yPos_.front(); }
+
+    /*!
+     * \brief Returns the maximum of the Y coordinate of the sampling points.
+     */
+    Scalar yMax() const
+    { return yPos_.back(); }
+
+    /*!
+     * \brief Returns the value of a sampling point.
+     */
+    Scalar valueAt(size_t i, size_t j) const
+    { return samples_[i][j]; }
+
+    /*!
+     * \brief Returns true if a coordinate lies in the tabulated range on the x direction
+     */
+    template <class Evaluation>
+    bool appliesX(const Evaluation& x) const
+    {
+        if (x < xMin() || xMax() < x) {
+            return false;
+        } else {
+            return true;
+        }
+    }
+
+    /*!
+     * \brief Returns true if a coordinate lies in the tabulated range on the y direction
+     */
+    template <class Evaluation>
+    bool appliesY(const Evaluation& y) const
+    {
+        if (y < yMin() || yMax() < y) {
+            return false;
+        } else {
+            return true;
+        }
+    }
+
+    /*!
+     * \brief Returns the number of sampling points in X direction.
+     */
+    size_t numX() const
+    { return xPos_.size(); }
+
+    /*!
+     * \brief Returns the number of sampling points in Y direction.
+     */
+    size_t numY() const
+    { return yPos_.size(); }
+
+    /*!
+     * \brief Return the interval index of a given position on the x-axis.
+     */
+    template <class Evaluation>
+    unsigned xSegmentIndex(const Evaluation& x) const
+    {
+        assert(xExtrapolate_ || (xMin() <= x && x <= xMax()));
+
+        // we need at least two sampling points!
+        assert(numX() >= 2);
+
+        if (x <= xMin() || numX() == 2)
+            return 0;
+        else if (x >= xMax())
+                return numX() - 2;
+        else {
+            assert(numX() >= 3);
+
+            // bisection
+            unsigned lowerIdx = 0;
+            unsigned upperIdx = numX() - 1;
+            while (lowerIdx + 1 < upperIdx) {
```

Comment:
Why not use std::lower_bound()?

Also: Logic is identical between x and y, use a single function, implement using common function (can keep the existing functions, and let them call a new common method).

Reply:
I changed to use `std::lower_bound()`. I have the impression that sometimes the meaning of some functions from algorithm is not `that` clear.